compiler/src/dotty/tools/dotc/transform/MacroAnnotations.scala

```diff
@@ -31,82 +31,100 @@ class MacroAnnotations(phase: IdentityDenotTransformer):
    *  Returns a list with transformed definition and any added definitions.
    */
-  def expandAnnotations(tree: MemberDef)(using Context): List[DefTree] =
+  def expandAnnotations(tree: MemberDef, companion: Option[MemberDef])(using Context): (List[MemberDef], Option[MemberDef]) =
     if !hasMacroAnnotation(tree.symbol) then
-      List(tree)
+      (List(tree), companion)
     else if tree.symbol.is(Module) && !tree.symbol.isClass then
       // only class is transformed
-      List(tree)
+      (List(tree), companion)
     else if tree.symbol.isType && !tree.symbol.isClass then
       report.error("macro annotations are not supported on type", tree)
-      List(tree)
+      (List(tree), companion)
     else
       debug.println(i"Expanding macro annotations of:\n$tree")
-
       val macroInterpreter = new Interpreter(tree.srcPos, MacroClassLoader.fromContext)
 
-      val allTrees = List.newBuilder[DefTree]
-      var insertedAfter: List[List[DefTree]] = Nil
+      val allTrees = List.newBuilder[MemberDef]
+      var insertedAfter: List[List[MemberDef]] = Nil
 
       // Apply all macro annotation to `tree` and collect new definitions in order
-      val transformedTree: DefTree = tree.symbol.annotations.foldLeft(tree) { (tree, annot) =>
-        if isMacroAnnotation(annot) then
-          debug.println(i"Expanding macro annotation: ${annot}")
-
-          // Interpret call to `new myAnnot(..).transform(using <Quotes>)(<tree>)`
-          val transformedTrees = callMacro(macroInterpreter, tree, annot)
-          transformedTrees.span(_.symbol != tree.symbol) match
-            case (prefixed, newTree :: suffixed) =>
-              allTrees ++= prefixed
-              insertedAfter = suffixed :: insertedAfter
-              for prefixedTree <- prefixed do
-                checkMacroDef(prefixedTree, tree, annot)
-              for suffixedTree <- suffixed do
-                checkMacroDef(suffixedTree, tree, annot)
-              TreeChecker.checkMacroGeneratedTree(tree, newTree)
-              newTree
-            case (Nil, Nil) =>
-              report.error(i"Unexpected `Nil` returned by `(${annot.tree}).transform(..)` during macro expansion", annot.tree.srcPos)
-              tree
-            case (_, Nil) =>
-              report.error(i"Transformed tree for ${tree} was not return by `(${annot.tree}).transform(..)` during macro expansion", annot.tree.srcPos)
-              tree
-        else
-          tree
-      }
-
+      val (transformedTree, transformedCompanion) =
+        tree.symbol.annotations.foldLeft((tree, companion)): (lastResult, annot) =>
+          val (tree, companion) = lastResult
+          if isMacroAnnotation(annot) then
+            debug.println(i"Expanding macro annotation: ${annot}")
+            // Interpret call to `new myAnnot(..).transform(using <Quotes>)(<tree>, <companion>)`
+            val (transformedTrees, transformedCompanion) = callMacro(macroInterpreter, tree, companion, annot)
```

Comment:
One issue that I still see in that design is the exapansion of the companion. What should we do when the companion was already processed but we now added a definition that needs inlining ? Should we rerun inlining ? We can't do that if there are MacroAnnotation as we might process them twice

Reply:
@nicolasstucki I forgot to discuss this case with you today (this might be a tricker one to resolve)